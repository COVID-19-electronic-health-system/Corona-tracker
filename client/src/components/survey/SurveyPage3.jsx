import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import { Typography, Button, TextField, Grid } from '@material-ui/core';
import { useBlockstack } from 'react-blockstack';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import buttonsCss from '../../css/buttons';
import actions from '../../redux/actions/actions';

const useStyles = makeStyles(theme => ({
  root: {
    alignItems: 'center',
    justifyContent: 'center',
    marginLeft: 'auto',
    marginRight: 'auto',
    maxWidth: theme.breakpoints.values.md,
    padding: '4px 32px',
  },
  additionalComments: {
    width: '100%',
    border: '2px solid #f64141',
  },
  continueButton: {
    ...buttonsCss.buttons,

    margin: '20px 8px 10px 8px',
    width: '160px',
  },
}));

const SurveyPage3 = props => {
  const { setSurveyPage3, survey, toSurveyPage2, submitSurvey } = props;
  const { nonPhysical } = survey;
  const classes = useStyles();
  const [openComment, setOpenComment] = useState(nonPhysical.openComment || '');
  const history = useHistory();
  const { userSession } = useBlockstack();

  const handleopenComment = value => {
    setOpenComment(value);
  };

  const sendBackToPage2 = () => {
    toSurveyPage2({ openComment });
  };

  const submitSurveyPage3 = async () => {
    survey.nonPhysical.openComment = openComment;

    submitSurvey(survey, userSession);

    history.push('/');

    setSurveyPage3({
      openComment,
    });

    window.localStorage.setItem('surveyCompleted', 'true');
  };

  return (
    <div className={classes.root}>
      <Grid container justify="center" spacing={1} className={classes.grid}>
        <Grid item xs={12}>
          <Typography>
            <b>Q6: Anything else you want to add?</b>
          </Typography>
        </Grid>
        <Grid item xs={12}>
          <TextField
            variant="outlined"
            className={classes.additionalComments}
            defaultValue={openComment}
            onChange={e => handleopenComment(e.target.value)}
          />
        </Grid>
        <Grid item xs={12}>
          <Button onClick={sendBackToPage2} variant="outlined" color="secondary" className={classes.continueButton}>
            BACK
          </Button>
          <Button onClick={submitSurveyPage3} color="secondary" className={classes.continueButton}>
            SUBMIT
          </Button>
        </Grid>
      </Grid>
    </div>
  );
};
SurveyPage3.propTypes = {
  setSurveyPage3: PropTypes.func.isRequired,
  survey: PropTypes.objectOf(Object).isRequired,
  toSurveyPage2: PropTypes.func.isRequired,
  submitSurvey: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  return {
    survey: state.surveyReducer.survey,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    setSurveyPage3: survey => dispatch(actions.setSurveyPage3(survey)),
    toSurveyPage2: survey => dispatch(actions.toSurveyPage2(survey)),
    submitSurvey: (fileNumber, observation, encryptOptions, userSession) =>
      dispatch(actions.submitSurveyThunk(fileNumber, observation, encryptOptions, userSession)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(SurveyPage3);
