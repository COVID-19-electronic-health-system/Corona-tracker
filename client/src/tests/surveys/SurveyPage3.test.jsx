import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { BrowserRouter } from 'react-router-dom';
import SurveyPage3 from '../../components/survey/SurveyPage3';

const mockStore = configureMockStore([thunk]);

const mockSurveyReducerState = {
  survey: {
    date: 0,
    physical: {},
    nonPhysical: {},
  },
  surveyPage: 3,
};

const mockObservationReducerState = {
  numObservations: 0,
};

describe('SurveyPage3', () => {
  it('should render a SurveyPage3 component ', () => {
    const store = mockStore({
      surveyReducer: mockSurveyReducerState,
      observationsReducer: mockObservationReducerState,
    });
    window.HTMLCanvasElement.prototype.getContext = () => {};

    const wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <SurveyPage3 />
        </Provider>
      </BrowserRouter>
    );
    expect(wrapper.find('SurveyPage3')).toMatchSnapshot();
  });
});

// TODO broken Redux call
// TODO make sure popup always pops up
