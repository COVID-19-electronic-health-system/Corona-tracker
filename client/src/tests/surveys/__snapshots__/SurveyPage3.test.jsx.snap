// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurveyPage3 should render a SurveyPage3 component  1`] = `ReactWrapper {}`;
